crates/webidl/src/lib.rs

```diff
@@ -396,6 +397,9 @@ impl<'src> FirstPassRecord<'src> {
             .to_syn_type(TypePosition::Argument)
             .unwrap_or(None)?;
 
+        let return_ty =
+            optional_return_ty(idl_type.to_syn_type(TypePosition::Return).unwrap().unwrap());
```

Comment:
> But it's annoying for us because it means there's no machine-readable way to tell, similar to `[Throws]`. I'm not sure if it'd be better to add a similar annotation here or just let everything be optional.

I think it would be nice to add an annotation for these use-cases, but we would have to do it now because it would be a breaking change later. Considering you already looked at this, do you think its feasible to quickly identify those dictionaries and mark them correctly or is it too big of a task?

> And there are instances of this, e.g. `MediaStreamContraints` uses fields with default values, which effectively means that they're optional when passed as an argument but required when returned.

Considering that we don't have two different types, one for use as an argument and a different one for use as a return value, I agree that in this case it just has to return `Option`.

---

I will go ahead and at least let required fields not return `Option`.

Reply:
> I will go ahead and at least let required fields not return `Option`.

Done.

Reply:
> I think it would be nice to add an annotation for these use-cases, but we would have to do it now because it would be a breaking change later. Considering you already looked at this, do you think its feasible to quickly identify those dictionaries and mark them correctly or is it too big of a task?

There aren't too many dictionaries that are used as return types, so I think it should be feasible; this is the full list:

- `AuthenticationExtensionsClientOutputs`
- `ComputedEffectTiming`
- `DOMQuadJSON`
- `DisplayNameResult`
- `EffectTiming`
- `FontFaceSetIteratorResult`
- `LocaleInfo`
- `MediaKeySystemConfiguration`
- `MediaStreamConstraints`
- `MediaTrackCapabilities`
- `MediaTrackConstraints`
- `MediaTrackSettings`
- `MediaTrackSupportedConstraints`
- `PushSubscriptionJSON`
- `RTCConfiguration`
- `RTCRtpParameters`
- `SerialPortInfo`
- `TreeCellInfo`
- `VideoColorSpaceInit`

<details>
<summary>I generated this by just adding some logging to `wasm-bindgen-webidl`:</summary>

```diff
diff --git a/crates/webidl/src/util.rs b/crates/webidl/src/util.rs
index 3e8cce3e4..94dddd51c 100644
--- a/crates/webidl/src/util.rs
+++ b/crates/webidl/src/util.rs
@@ -330,6 +330,9 @@ impl<'src> FirstPassRecord<'src> {
             //       do this much earlier to avoid all the above work if
             //       possible.
             let ret_ty = signature.orig.ret.to_idl_type(self);
+            if let IdlType::Dictionary(name) = ret_ty {
+                println!("{name} used in return position")
+            }
 
             let mut rust_name = snake_case_ident(name);
             let mut first = true;
```
</details>

Reply:
Done!

Also deprecated/removed their constructors, e.g. otherwise you can just write `EffectTiming::new().get_fill()` and it would fail.

Reply:
Actually, I just realized, that there should be a ton of dictionaries returned from `readonly attribute`s that would fall into this category as well. Additionally, dictionaries can be wrapped in `Promise` or `sequence` that we didn't consider here as well.

Ideally we should programmatically detect dictionaries that are never used as a parameter?

In any case, I'm going to revisit this again tomorrow.

Reply:
Just saw in CI that we can't deprecate `EffectTiming::default()`, so that's also less then ideal.

Reply:
See https://github.com/daxpedda/wasm-bindgen/commit/0d4bd7ef259c54573d0b9233b13f9a91123b4858.
I believe that should be pretty accurate from what we can extract from the WebIDL we have.

<details>
  <summary>List of Dictionaries</summary>
    
  ```
  DOMQuadJSON
  EcdsaParams
  ClientRectsAndTexts
  WebTransportDatagramStats
  DecoderDoctorNotification
  IntersectionObserverEntryInit
  DhImportKeyParams
  RTCMediaStreamStats
  PermissionDescriptor
  Transformer
  RTCIdentityAssertion
  LockManagerSnapshot
  ConsoleCounterError
  BasePropertyIndexedKeyframe
  FontFaceSetIteratorResult
  CollectedClientData
  NetworkResultOptions
  FakePluginTagInit
  RTCIceComponentStats
  ReadableStreamReadResult
  Pbkdf2Params
  MediaTrackSettings
  ConsoleTimerStart
  RsaOtherPrimesInfo
  HIDCollectionInfo
  DOMWindowResizeEventDetail
  ProfileTimelineMarker
  DhKeyGenParams
  ImageDecodeResult
  RsaHashedKeyGenParams
  SerialPortInfo
  NavigatorUABrandVersion
  MemoryMeasurement
  RequestMediaKeySystemAccessNotification
  CheckerboardReport
  LocaleInfo
  RsaPssParams
  VideoDecoderSupport
  TreeCellInfo
  AesCbcParams
  RsaHashedImportParams
  MemoryAttributionContainer
  IterableKeyOrValueResult
  ConsoleProfileEvent
  RTCIceCandidateStats
  MediaTrackSupportedConstraints
  DNSCacheDict
  FetchReadableStreamReadDataArray
  RTCRTPStreamStats
  BasicCardRequest
  MemoryBreakdownEntry
  WidevineCDMManifest
  RTCRTPContributingSourceStats
  XRPermissionDescriptor
  DoubleRange
  KeyIdsInitData
  ConnStatusDict
  VideoEncoderSupport
  DisplayNameResult
  ProfileTimelineLayerRect
  RsaHashedKeyAlgorithm
  RTCRtpCapabilities
  ConsoleCounter
  SocketElement
  WebTransportStats
  NativeOSFileWriteAtomicOptions
  AesCtrParams
  BasicCardResponse
  HttpConnDict
  DNSLookupDict
  SocketsDict
  BaseKeyframe
  LockInfo
  EncodedVideoChunkMetadata
  ProfileTimelineStackFrame
  SignResponse
  RTCCodecStats
  PushSubscriptionJSON
  BaseComputedKeyframe
  HIDReportInfo
  AesGcmParams
  EcKeyImportParams
  AudioEncoderSupport
  WebTransportReceiveStreamStats
  U2FClientData
  FetchReadableStreamReadDataDone
  ULongRange
  DnsCacheEntry
  ClipboardPermissionDescriptor
  UALowEntropyJSON
  FakePluginMimeEntry
  WebGLContextAttributes
  XRSessionSupportedPermissionDescriptor
  MutationObservingInfo
  RTCCertificateExpiration
  MediaTrackCapabilities
  RTCRtpSynchronizationSource
  SerialInputSignals
  IterableKeyAndValueResult
  AnimationPropertyDetails
  HkdfParams
  RTCRtpHeaderExtensionCapability
  JsonWebKey
  EncodedAudioChunkMetadata
  BluetoothDataFilterInit
  ConsoleTimerError
  RTCStatsReportInternal
  CryptoKeyPair
  SvcOutputMetadata
  AesDerivedKeyParams
  NetworkCommandOptions
  HttpConnectionElement
  EcKeyGenParams
  WebTransportSendStreamStats
  OpenWindowEventDetail
  KeyAlgorithm
  HttpConnInfo
  HmacImportParams
  BluetoothPermissionStorage
  DhKeyAlgorithm
  HmacKeyAlgorithm
  HIDReportItem
  LifecycleCallbacks
  MediaSessionActionDetails
  DOMQuadInit
  DOMMatrixInit
  AuthenticationExtensionsClientOutputs
  RsaOaepParams
  MemoryAttribution
  HalfOpenInfoDict
  CacheBatchOperation
  AesKeyAlgorithm
  RTCTransportStats
  EcKeyAlgorithm
  RcwnStatus
  RTCRtpContributingSource
  ConsoleTimerLogOrEnd
  DhKeyDeriveParams
  USBPermissionDescriptor
  ChromeFilePropertyBag
  EcdhKeyDeriveParams
  AllowedBluetoothDevice
  UADataValues
  RcwnPerfStats
  RTCOutboundRTPStreamStats
  HmacKeyGenParams
  RTCIceCandidatePairStats
  L10nElement
  WebSocketElement
  ContextAttributes2D
  AudioDecoderSupport
  BluetoothPermissionDescriptor
  AutocompleteInfo
  RegisterResponse
  RTCRtpSourceEntry
  RTCInboundRTPStreamStats
  ConsoleEvent
  ConsoleStackEntry
  WebSocketDict
  ComputedEffectTiming
  PushSubscriptionKeys
  RTCMediaStreamTrackStats
  USBPermissionStorage
  HmacDerivedKeyParams
  RTCStats
  UnderlyingSource
  UnderlyingSink
  NativeOSFileReadOptions
  Algorithm
  BlockParsingOptions
  FileSystemPermissionDescriptor
  AesKeyGenParams
  AnimationPropertyValueDetails
  StorageEstimate
  AllowedUSBDevice
  ```
</details>

175 entries, I found a lot of APIs there that were just copied from Firefox but aren't probably real accessible Web APIs, its still a lot to go through.
As I mentioned in https://github.com/rustwasm/wasm-bindgen/pull/3993#discussion_r1667735460, I believe we should try and drop this improvement entirely, unfortunately.

I guess if somebody wants to do the legwork here, what needs to be done is to go through each one of those and go to the spec and find out if really all members are always present or which ones aren't so we can mark them, e.g. like in `MediaTrackSettings`.